File: src/ComputeDeltaH.cpp
#include "..\include\itm.h"

/*=============================================================================
 |
 |  Description:  Compute the terrain irregularity parameter, delta_h
 |
 |        Input:  pfl[]          - Terrain data
 |                d_start__meter - Distance into the terrain profile to start 
 |                                 considering data, in meters
 |                d_end__meter   - Distance into the terrain profile to end 
 |                                 considering data, in meters
 |
 |      Outputs:  [None]
 |
 |      Returns:  delta_h__meter - Terrain irregularity parameter, in meters
 |
 *===========================================================================*/
double ComputeDeltaH(double pfl[], double d_start__meter, double d_end__meter)
{
    double s[247] = { 0 };                      // Temp pfl data array

    int np = (int)pfl[0];
    double x_start = d_start__meter / pfl[1];   // index to start considering terrain points
    double x_end = d_end__meter / pfl[1];       // index to stop considering terrain points

    // if there are less than 2 terrain points, return delta_h = 0
    if (x_end - x_start < 2.0)
        return 0;

    int p10 = (int)(0.1 * (x_end - x_start + 8.0));
    p10 = MIN(MAX(4, p10), 25);                 // 10% index

    int n = 10 * p10 - 5;
    int p90 = n - p10;                         // 90% index

    double np_s = n - 1;
    s[0] = np_s;
    s[1] = 1.0;

    x_end = (x_end - x_start) / np_s;
    int i = (int)x_start;
    x_start -= float(i + 1.0);

    for (int j = 0; j < n; j++)
    {
        while (x_start > 0.0 && (i + 1) < np)
        {
            x_start--;
            i++;
        }

        s[j + 2] = pfl[i + 3] + (pfl[i + 3] - pfl[i + 2]) * x_start;

        x_start += x_end;
    }

    double fit_y1, fit_y2;
    LinearLeastSquaresFit(s, 0.0, np_s, &fit_y1, &fit_y2);

    fit_y2 = (fit_y2 - fit_y1) / np_s;

    std::vector<double> diffs;
    
    // compute the difference between fitted line and actual data
    for (int j = 0; j < n; j++)
    {
        diffs.push_back(s[j + 2] - fit_y1);

        fit_y1 += fit_y2;
    }

    std::nth_element(diffs.begin(), diffs.begin() + p10 - 1, diffs.end(), std::greater<double>());
    double q10 = diffs[p10 - 1];

    std::nth_element(diffs.begin(), diffs.begin() + p90, diffs.end(), std::greater<double>());
    double q90 = diffs[p90];

    double delta_h_d__meter = q10 - q90;

    // [ERL 79-ITS 67, Eqn 3], inverted
    double delta_h__meter = delta_h_d__meter / (1.0 - 0.8 * exp(-(d_end__meter - d_start__meter) / 50e3));

    return delta_h__meter;
}

File: itm-rs/itm/src/compute_delta_h.rs
use crate::linear_least_squares_fit::linear_least_squares_fit;

/// Compute the terrain irregularity parameter delta_h between
/// `d_start__meter` and `d_end__meter` in the terrain profile `pfl`.
pub fn compute_delta_h(pfl: &[f64], d_start__meter: f64, d_end__meter: f64) -> f64 {
    let np = pfl[0] as usize;
    let x_start_f = d_start__meter / pfl[1];
    let x_end_f = d_end__meter / pfl[1];

    if x_end_f - x_start_f < 2.0 {
        return 0.0;
    }

    let p10_raw = (0.1 * (x_end_f - x_start_f + 8.0)) as usize;
    let p10 = p10_raw.clamp(4, 25);
    let n = 10 * p10 - 5;
    let p90 = n - p10;

    let np_s = (n - 1) as f64;
    let mut s = vec![0.0f64; n + 2];
    s[0] = np_s;
    s[1] = 1.0;

    let x_step = (x_end_f - x_start_f) / np_s;
    let mut i = x_start_f as usize;
    let mut x_start = x_start_f - (i + 1) as f64;

    for j in 0..n {
        while x_start > 0.0 && i + 1 < np {
            x_start -= 1.0;
            i += 1;
        }
        s[j + 2] = pfl[i + 3] + (pfl[i + 3] - pfl[i + 2]) * x_start;
        x_start += x_step;
    }

    let (fit_y1, fit_y2_raw) = linear_least_squares_fit(&s, 0.0, np_s);
    let fit_step = (fit_y2_raw - fit_y1) / np_s;

    let mut diffs: Vec<f64> = Vec::with_capacity(n);
    let mut running = fit_y1;
    for j in 0..n {
        diffs.push(s[j + 2] - running);
        running += fit_step;
    }

    // 10th percentile (largest) and 90th percentile (smallest of remaining)
    diffs.select_nth_unstable_by(p10 - 1, |a, b| b.partial_cmp(a).unwrap());
    let q10 = diffs[p10 - 1];
    diffs.select_nth_unstable_by(p90, |a, b| b.partial_cmp(a).unwrap());
    let q90 = diffs[p90];

    let delta_h_d__meter = q10 - q90;
    delta_h_d__meter / (1.0 - 0.8 * (-(d_end__meter - d_start__meter) / 50_000.0).exp())
}
